Forge.TreeWalker/src/TreeSchemaValidator.cs

```diff
@@ -0,0 +1,275 @@
+                success = true;
+                return schemaList;
+            }
+            catch
+            {
+                success = false;
+                return new List<object>();
+            }
+        }
+        private Task<Tuple<bool, IList<string>>> Validate(List<object> schemas, object rules)
+        {
+            var errorList = new List<string>();
+            try
+            {
+                JSchema jRules;
+                //input rules could be string, Jschema type. Null means using Default rules
+                if (rules is null)
+                    jRules = this.ForgeSchemaValidationRules;
+                else if (rules is JSchema)
+                    jRules = (JSchema)rules;
+                else
+                    jRules = JSchema.Parse((string)rules);
+
+                var schemaList = new List<JObject>();
+                foreach (var item in schemas)
+                {
+                    //schmeas type could be string or forgeTree
+                    if (item is string)
+                    {
+
+                        Dictionary<string, ForgeTree> forgeTrees = JsonConvert.DeserializeObject<Dictionary<string, ForgeTree>>((string)item);
+                        foreach (var kvp in forgeTrees)
+                        {
+                            ForgeTree forgeTree = kvp.Value;
+                            if (forgeTree.Tree == null)
+                            {
+                                // Deserialize into Dictionary does not throw exception but will have null "Tree" property if schema is just a ForgeTree.
+                                // try to deserialize string to forge tree directly
+                                JsonConvert.DeserializeObject<ForgeTree>((string)item);
+                                JObject schema = JObject.Parse((string)item);
+                                schemaList.Add(schema);
+                                break;
+                            }
+                            SerializeForgeTree(schemaList, forgeTree);
+                        }
+
+                    }
+                    else
+                    {
+                        SerializeForgeTree(schemaList, (ForgeTree)item);
+                    }
+                }
+                if (schemaList.Count == 0)
+                {
+                    errorList.Add("Can't find target schema to test or file type is not supported");
+                    return Task.FromResult(new Tuple<bool, IList<string>>(false, errorList));
+                }
+                foreach (var item in schemaList)
+                {
+                    var validated = DoValidate(item, jRules, out IList<ValidationError> errors);
+                    if (!validated)
+                    {
+                        foreach (var error in errors)
+                        {
+                            errorList.Add(error.Message + " line: " + error.LineNumber + " position: " + error.LinePosition);
```

Comment:
>error [](http://example.com/codeflow?start=42&length=5)

What does just printing the error look like? If it includes the line number and position already, we can probably just use that. Can you paste an example?

Reply:
Were you able to get an example what this message looks like?

Reply:
Still wondering about this. I'd prefer to just propagate up the errorList from IsValid unless there is a good reason to change it.